File: chad/src/main/java/org/woahoverflow/chad/core/ChadVar.java
package org.woahoverflow.chad.core;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.source.soundcloud.SoundCloudAudioSourceManager;
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager;
import org.woahoverflow.chad.commands.admin.*;
import org.woahoverflow.chad.commands.community.*;
import org.woahoverflow.chad.commands.developer.Shutdown;
import org.woahoverflow.chad.commands.developer.*;
import org.woahoverflow.chad.commands.fun.*;
import org.woahoverflow.chad.commands.gambling.Balance;
import org.woahoverflow.chad.commands.gambling.CoinFlip;
import org.woahoverflow.chad.commands.gambling.DailyReward;
import org.woahoverflow.chad.commands.info.*;
import org.woahoverflow.chad.commands.music.*;
import org.woahoverflow.chad.commands.nsfw.Hentai;
import org.woahoverflow.chad.commands.nsfw.Porn;
import org.woahoverflow.chad.commands.punishments.Ban;
import org.woahoverflow.chad.commands.punishments.Kick;
import org.woahoverflow.chad.framework.handle.JsonHandler;
import org.woahoverflow.chad.framework.obj.Command;
import org.woahoverflow.chad.framework.obj.Command.Category;
import org.woahoverflow.chad.framework.obj.Command.Data;
import org.woahoverflow.chad.framework.obj.GuildMusicManager;
import sx.blah.discord.handle.obj.StatusType;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Storing of Static Variables that don't have a home
 *
 * @author sho
 */
public final class ChadVar {
    public static final String VERSION = "v0.9.4";

    /**
     * Gigantic Words List
     */
    public static List<String> wordsList = new ArrayList<>();

    /**
     * All the swear words for the swear filter
     */
    public static final List<String> swearWords = new ArrayList<>();

    /**
     * All results available for the eight ball command
     */
    public static final List<String> eightBallResults = new ArrayList<>();

    /**
     * The universal player manager for music playing
     */
    public static final AudioPlayerManager playerManager = new DefaultAudioPlayerManager();

    /**
     * The music manager for guilds
     */
    public static final ConcurrentHashMap<Long, GuildMusicManager> musicManagers = new ConcurrentHashMap<>();

    /**
     * The Youtube API Key in the bot.json file
     */
    public static String YOUTUBE_API_KEY = JsonHandler.handle.get("youtube_api_key");

    /**
     * The Steam API key in the bot.json file
     */
    public static String STEAM_API_KEY = JsonHandler.handle.get("steam_api_key");

    /*
      Registers sources for the player manager
     */
    static {
        AudioSourceManagers.registerRemoteSources(playerManager);
        playerManager.registerSourceManager(new YoutubeAudioSourceManager());
        playerManager.registerSourceManager(new SoundCloudAudioSourceManager());
    }

    /**
     * The status type (idle, online, offline, dnd)
     */
    public static StatusType statusType = StatusType.ONLINE;

    /**
     * The current status string
     */
    public static String currentStatus = "";

    /**
     * The amount of time to take to rotate between presences
     */
    public static final int rotationInteger = 60*5; // 5 minutes

    /**
     * If it should rotate presence at all
     */
    public static boolean rotatePresence = true;

    /**
     * The presences to rotate between
     */
    public static final List<String> presenceRotation = new ArrayList<>();

    /**
     * The full list of Commands
     */
    public static final ConcurrentHashMap<String, Command.Data> COMMANDS = new ConcurrentHashMap<>();

    /**
     * List of Verified Developers
     */
    public static final List<Long> DEVELOPERS = new ArrayList<>();

    /*
      Define all of the Commands
     */
    static {
        // FUN!
        COMMANDS.put("random", new Command.Data(Command.Category.FUN, new Random()));
        COMMANDS.put("photoeditor", new Command.Data(Command.Category.FUN, new PhotoEditor(), "pe"));
        COMMANDS.put("eightball", new Command.Data(Command.Category.FUN, new EightBall(), "8ball"));
        COMMANDS.put("catgallery", new Command.Data(Command.Category.FUN, new CatGallery(), "catgal"));
        COMMANDS.put("catfact", new Command.Data(Command.Category.FUN, new CatFact()));
        COMMANDS.put("russianroulette", new Command.Data(Command.Category.FUN, new RussianRoulette(), "rrl"));
        COMMANDS.put("wordreverse", new Command.Data(Command.Category.FUN, new WordReverse(), "wr"));
        COMMANDS.put("rockpaperscissors", new Command.Data(Command.Category.FUN, new RockPaperScissors(), "rps"));
        COMMANDS.put("meme", new Data(Category.FUN, new Meme()));
        COMMANDS.put("doggallery", new Data(Category.FUN, new DogGallery()));
        COMMANDS.put("dogfact", new Data(Category.FUN, new DogFact()));

        // COMMUNITY!
        COMMANDS.put("divorce", new Data(Category.COMMUNITY, new DivorcePlayer()));
        COMMANDS.put("marry", new Data(Category.COMMUNITY, new MarryPlayer()));
        COMMANDS.put("cuddle", new Data(Category.COMMUNITY, new Cuddle()));
        COMMANDS.put("profile", new Data(Category.COMMUNITY, new Profile()));
        COMMANDS.put("upvote", new Data(Category.COMMUNITY, new UpVote()));
        COMMANDS.put("downvote", new Data(Category.COMMUNITY, new DownVote()));

        // INFO!
        COMMANDS.put("help", new Command.Data(Command.Category.INFO, new Help()));
        COMMANDS.put("userinfo", new Command.Data(Command.Category.INFO, new UserInfo(), "uinfo"));
        COMMANDS.put("steam", new Command.Data(Command.Category.INFO, new Steam(), "st"));
        COMMANDS.put("chad", new Command.Data(Command.Category.INFO, new Chad()));
        COMMANDS.put("guildinfo", new Command.Data(Command.Category.INFO, new GuildInfo(), "ginfo"));
        COMMANDS.put("reddittop", new Command.Data(Command.Category.INFO, new RedditTop(), "rtop"));
        COMMANDS.put("redditnew", new Command.Data(Command.Category.INFO, new RedditNew(), "rnew"));
        COMMANDS.put("subscribercount", new Data(Category.INFO, new SubscriberCount(), "subcount", "subc"));
        COMMANDS.put("steamstatus", new Data(Category.INFO, new SteamStatus(), "steamst"));
        COMMANDS.put("aliases", new Data(Category.INFO, new Aliases(), "alias"));

        // PUNISHMENTS!
        COMMANDS.put("kick", new Command.Data(Command.Category.PUNISHMENTS, new Kick()));
        COMMANDS.put("ban", new Command.Data(Command.Category.PUNISHMENTS, new Ban()));

        // ADMINISTRATOR!
        COMMANDS.put("prefix", new Command.Data(Command.Category.ADMINISTRATOR, new Prefix()));
        COMMANDS.put("logging", new Command.Data(Command.Category.ADMINISTRATOR, new Logging()));
        COMMANDS.put("purge", new Command.Data(Command.Category.ADMINISTRATOR, new Purge()));
        COMMANDS.put("instantmessage", new Command.Data(Command.Category.ADMINISTRATOR, new Message(), "im"));
        COMMANDS.put("autorole", new Command.Data(Command.Category.ADMINISTRATOR, new AutoRole(), "ar"));
        COMMANDS.put("permissions", new Command.Data(Command.Category.ADMINISTRATOR, new Permissions(), "perms"));
        COMMANDS.put("guildsettings", new Data(Category.ADMINISTRATOR, new GuildSettings(), "gset"));
        COMMANDS.put("nsfw", new Command.Data(Command.Category.ADMINISTRATOR, new Nsfw()));
        COMMANDS.put("swearfilter", new Command.Data(Category.ADMINISTRATOR, new Swearing(), "sf"));

        // NSFW !
        COMMANDS.put("porn", new Command.Data(Command.Category.NSFW, new Porn(), "pn"));
        COMMANDS.put("hentai", new Command.Data(Command.Category.NSFW, new Hentai(), "hentie"));

        // DEVELOPER!
        COMMANDS.put("threads", new Command.Data(Command.Category.DEVELOPER, new CurrentThreads(), "cth"));
        COMMANDS.put("modpresence", new Command.Data(Command.Category.DEVELOPER, new ModifyPresence(), "modp"));
        COMMANDS.put("systeminfo", new Command.Data(Command.Category.DEVELOPER, new SystemInfo(), "sinf"));
        COMMANDS.put("shutdown", new Command.Data(Command.Category.DEVELOPER,new Shutdown()));
        COMMANDS.put("modifybalance", new Command.Data(Command.Category.DEVELOPER, new ModifyBalance(), "modbal"));
        COMMANDS.put("modifycache", new Data(Category.DEVELOPER, new ModifyCache(), "modcache"));
        COMMANDS.put("modifydatabase", new Data(Category.DEVELOPER, new ModifyDatabase(), "moddb"));
        COMMANDS.put("modifydevelopers", new Data(Category.DEVELOPER, new ModifyDevelopers(), "moddev"));
        COMMANDS.put("sync", new Data(Category.DEVELOPER, new Sync()));

        // GAMBLING!
        COMMANDS.put("coinflip", new Command.Data(Command.Category.GAMBLING, new CoinFlip(), "cf"));
        COMMANDS.put("balance", new Command.Data(Command.Category.GAMBLING, new Balance(), "bal"));
        COMMANDS.put("dailyreward", new Data(Category.GAMBLING, new DailyReward(), "drw"));

        // MUSIC
        COMMANDS.put("play", new Data(Category.MUSIC, new Play(), "p"));
        COMMANDS.put("pause", new Data(Category.MUSIC, new Pause(), "pp"));
        COMMANDS.put("leave", new Data(Category.MUSIC, new Leave(), "ll"));
        COMMANDS.put("skip", new Data(Category.MUSIC, new Skip(), "ss"));
        COMMANDS.put("queue", new Data(Category.MUSIC, new Queue(), "qq"));
        COMMANDS.put("volume", new Data(Category.MUSIC, new Volume(), "vol"));
        COMMANDS.put("resume", new Data(Category.MUSIC, new Resume(), "res", "rr"));
    }
}

File: chad/src/main/java/org/woahoverflow/chad/core/ChadVar.kt
package org.woahoverflow.chad.core

import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.source.soundcloud.SoundCloudAudioSourceManager
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager
import org.woahoverflow.chad.commands.Hentai
import org.woahoverflow.chad.commands.Meme
import org.woahoverflow.chad.commands.Porn
import org.woahoverflow.chad.commands.`fun`.*
import org.woahoverflow.chad.commands.admin.*
import org.woahoverflow.chad.commands.community.*
import org.woahoverflow.chad.commands.developer.*
import org.woahoverflow.chad.commands.gambling.Balance
import org.woahoverflow.chad.commands.gambling.CoinFlip
import org.woahoverflow.chad.commands.gambling.DailyReward
import org.woahoverflow.chad.commands.info.*
import org.woahoverflow.chad.commands.music.*
import org.woahoverflow.chad.commands.punishments.Ban
import org.woahoverflow.chad.commands.punishments.Kick
import org.woahoverflow.chad.commands.punishments.Mute
import org.woahoverflow.chad.framework.handle.JsonHandler
import org.woahoverflow.chad.framework.obj.Command
import org.woahoverflow.chad.framework.obj.Command.Data
import org.woahoverflow.chad.framework.obj.GuildMusicManager
import sx.blah.discord.handle.obj.ActivityType
import sx.blah.discord.handle.obj.StatusType
import java.util.concurrent.ConcurrentHashMap

/**
 * Storing of Static Variables that don't have a home
 *
 * @author sho
 */
object ChadVar {
    const val VERSION = "v1.0.0"

    /**
     * Gigantic Words List
     */
    @JvmStatic
    var wordsList: ArrayList<String> = arrayListOf()

    /**
     * All the swear words for the swear filter
     */
    @JvmStatic
    val swearWords: ArrayList<String> = arrayListOf()

    /**
     * All results available for the eight ball command
     */
    @JvmStatic
    val eightBallResults: ArrayList<String> = arrayListOf()

    /**
     * The universal player manager for music playing
     */
    @JvmStatic
    val playerManager: AudioPlayerManager = DefaultAudioPlayerManager()

    /**
     * The music manager for guilds
     */
    @JvmStatic
    val musicManagers = ConcurrentHashMap<Long, GuildMusicManager>()

    /**
     * The Youtube API Key in the bot.json file
     */
    @JvmStatic
    var YOUTUBE_API_KEY = JsonHandler["youtube_api_key"]

    /**
     * The Steam API key in the bot.json file
     */
    @JvmStatic
    var STEAM_API_KEY = JsonHandler["steam_api_key"]

    /**
     * The status type (idle, online, offline, dnd)
     */
    @JvmStatic
    var statusType = StatusType.ONLINE

    /**
     * The activity type
     */
    @JvmStatic
    var activityType = ActivityType.PLAYING

    /**
     * The current status string
     */
    @JvmStatic
    var currentStatus = ""

    /**
     * If it should rotate presence at all
     */
    @JvmStatic
    var rotatePresence = true

    /**
     * The full list of Commands
     */
    @JvmStatic
    val COMMANDS = object : ConcurrentHashMap<String, Data>() {
        init {
            // FUN!
            this["random"] = Data(Command.Category.FUN, Random())
            this["photoeditor"] = Data(Command.Category.FUN, PhotoEditor(), "pe")
            this["eightball"] = Data(Command.Category.FUN, EightBall(), "8ball")
            this["catgallery"] = Data(Command.Category.FUN, CatGallery(), "catgal")
            this["catfact"] = Data(Command.Category.FUN, CatFact())
            this["russianroulette"] = Data(Command.Category.FUN, RussianRoulette(), "rrl")
            this["wordreverse"] = Data(Command.Category.FUN, WordReverse(), "wr")
            this["rockpaperscissors"] = Data(Command.Category.FUN, RockPaperScissors(), "rps")
            this["meme"] = Data(Command.Category.FUN, Meme())
            this["doggallery"] = Data(Command.Category.FUN, DogGallery())
            this["dogfact"] = Data(Command.Category.FUN, DogFact())

            // COMMUNITY!
            this["divorce"] = Data(Command.Category.COMMUNITY, DivorcePlayer())
            this["marry"] = Data(Command.Category.COMMUNITY, MarryPlayer())
            this["cuddle"] = Data(Command.Category.COMMUNITY, Cuddle())
            this["profile"] = Data(Command.Category.COMMUNITY, Profile())
            this["upvote"] = Data(Command.Category.COMMUNITY, UpVote())
            this["downvote"] = Data(Command.Category.COMMUNITY, DownVote())

            // INFO!
            this["help"] = Data(Command.Category.INFO, Help())
            this["userinfo"] = Data(Command.Category.INFO, UserInfo(), "uinfo")
            this["steam"] = Data(Command.Category.INFO, Steam(), "st")
            this["chad"] = Data(Command.Category.INFO, Chad())
            this["guildinfo"] = Data(Command.Category.INFO, GuildInfo(), "ginfo")
            this["reddittop"] = Data(Command.Category.INFO, RedditTop(), "rtop")
            this["redditnew"] = Data(Command.Category.INFO, RedditNew(), "rnew")
            this["subscribercount"] = Data(Command.Category.INFO, SubscriberCount(), "subcount", "subc")
            this["steamstatus"] = Data(Command.Category.INFO, SteamStatus(), "steamst")
            this["aliases"] = Data(Command.Category.INFO, Aliases(), "alias")
            this["leaderboard"] = Data(Command.Category.INFO, Leaderboard(), "lb")

            // PUNISHMENTS!
            this["kick"] = Data(Command.Category.PUNISHMENTS, Kick())
            this["ban"] = Data(Command.Category.PUNISHMENTS, Ban())
            this["mute"] = Data(Command.Category.PUNISHMENTS, Mute())

            // ADMINISTRATOR!
            this["prefix"] = Data(Command.Category.ADMINISTRATOR, Prefix())
            this["logging"] = Data(Command.Category.ADMINISTRATOR, Logging())
            this["purge"] = Data(Command.Category.ADMINISTRATOR, Purge())
            this["instantmessage"] = Data(Command.Category.ADMINISTRATOR, Message(), "im")
            this["autorole"] = Data(Command.Category.ADMINISTRATOR, AutoRole(), "ar")
            this["permissions"] = Data(Command.Category.ADMINISTRATOR, Permissions(), "perms")
            this["guildsettings"] = Data(Command.Category.ADMINISTRATOR, GuildSettings(), "gset")
            this["nsfw"] = Data(Command.Category.ADMINISTRATOR, Nsfw())
            this["swearfilter"] = Data(Command.Category.ADMINISTRATOR, Swearing(), "sf")

            // NSFW !
            this["porn"] = Data(Command.Category.NSFW, Porn(), "pn")
            this["hentai"] = Data(Command.Category.NSFW, Hentai(), "hentie")

            // DEVELOPER!
            this["threads"] = Data(Command.Category.DEVELOPER, CurrentThreads(), "cth")
            this["modpresence"] = Data(Command.Category.DEVELOPER, ModifyPresence(), "modp")
            this["systeminfo"] = Data(Command.Category.DEVELOPER, SystemInfo(), "sinf")
            this["shutdown"] = Data(Command.Category.DEVELOPER, Shutdown())
            this["modifybalance"] = Data(Command.Category.DEVELOPER, ModifyBalance(), "modbal")
            this["modifycache"] = Data(Command.Category.DEVELOPER, ModifyCache(), "modcache")
            this["modifydatabase"] = Data(Command.Category.DEVELOPER, ModifyDatabase(), "moddb")
            this["modifydevelopers"] = Data(Command.Category.DEVELOPER, ModifyDevelopers(), "moddev")
            this["sync"] = Data(Command.Category.DEVELOPER, Sync())
            this["statistics"] = Data(Command.Category.DEVELOPER, Statistics())

            // GAMBLING!
            this["coinflip"] = Data(Command.Category.GAMBLING, CoinFlip(), "cf")
            this["balance"] = Data(Command.Category.GAMBLING, Balance(), "bal")
            this["dailyreward"] = Data(Command.Category.GAMBLING, DailyReward(), "drw")

            // MUSIC
            this["play"] = Data(Command.Category.MUSIC, Play(), "p")
            this["pause"] = Data(Command.Category.MUSIC, Pause(), "pp")
            this["leave"] = Data(Command.Category.MUSIC, Leave(), "ll")
            this["skip"] = Data(Command.Category.MUSIC, Skip(), "ss")
            this["queue"] = Data(Command.Category.MUSIC, Queue(), "qq")
            this["volume"] = Data(Command.Category.MUSIC, Volume(), "vol")
            this["resume"] = Data(Command.Category.MUSIC, Resume(), "res")
        }
    }

    /**
     * List of Verified Developers
     */
    @JvmStatic
    val DEVELOPERS: ArrayList<Long> = arrayListOf()

    /**
     * The originally cached developers.
     *
     * Adds a useful fallback if an original developer gets removed through mistake etc.
     */
    @JvmStatic
    val ORIGINAL_DEVELOPERS: ArrayList<Long> = arrayListOf()

    /*
      Registers sources for the player manager
     */
    init {
        AudioSourceManagers.registerRemoteSources(playerManager)
        playerManager.registerSourceManager(YoutubeAudioSourceManager())
        playerManager.registerSourceManager(SoundCloudAudioSourceManager())
    }
}
